File: tests/unit_tests/commands/export_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=invalid-name, unused-argument, import-outside-toplevel

from unittest.mock import patch

import pytest
import yaml
from freezegun import freeze_time
from pytest_mock import MockerFixture

from superset.extensions import feature_flag_manager


def test_export_assets_command(mocker: MockerFixture) -> None:
    """
    Test that all assets are exported correctly.
    """
    from superset.commands.export.assets import ExportAssetsCommand

    ExportDatabasesCommand = mocker.patch(  # noqa: N806
        "superset.commands.export.assets.ExportDatabasesCommand"
    )
    ExportDatabasesCommand.return_value.run.return_value = [
        (
            "metadata.yaml",
            lambda: "version: 1.0.0\ntype: Database\ntimestamp: '2022-01-01T00:00:00+00:00'\n",  # noqa: E501
        ),
        ("databases/example.yaml", lambda: "<DATABASE CONTENTS>"),
    ]
    ExportDatasetsCommand = mocker.patch(  # noqa: N806
        "superset.commands.export.assets.ExportDatasetsCommand"
    )
    ExportDatasetsCommand.return_value.run.return_value = [
        (
            "metadata.yaml",
            lambda: "version: 1.0.0\ntype: Dataset\ntimestamp: '2022-01-01T00:00:00+00:00'\n",  # noqa: E501
        ),
        ("datasets/example/dataset.yaml", lambda: "<DATASET CONTENTS>"),
    ]
    ExportChartsCommand = mocker.patch(  # noqa: N806
        "superset.commands.export.assets.ExportChartsCommand"
    )
    ExportChartsCommand.return_value.run.return_value = [
        (
            "metadata.yaml",
            lambda: "version: 1.0.0\ntype: Slice\ntimestamp: '2022-01-01T00:00:00+00:00'\n",  # noqa: E501
        ),
        ("charts/pie.yaml", lambda: "<CHART CONTENTS>"),
    ]
    ExportDashboardsCommand = mocker.patch(  # noqa: N806
        "superset.commands.export.assets.ExportDashboardsCommand"
    )
    ExportDashboardsCommand.return_value.run.return_value = [
        (
            "metadata.yaml",
            lambda: "version: 1.0.0\ntype: Dashboard\ntimestamp: '2022-01-01T00:00:00+00:00'\n",  # noqa: E501
        ),
        ("dashboards/sales.yaml", lambda: "<DASHBOARD CONTENTS>"),
    ]
    ExportSavedQueriesCommand = mocker.patch(  # noqa: N806
        "superset.commands.export.assets.ExportSavedQueriesCommand"
    )
    ExportSavedQueriesCommand.return_value.run.return_value = [
        (
            "metadata.yaml",
            lambda: "version: 1.0.0\ntype: SavedQuery\ntimestamp: '2022-01-01T00:00:00+00:00'\n",  # noqa: E501
        ),
        ("queries/example/metric.yaml", lambda: "<SAVED QUERY CONTENTS>"),
    ]

    with freeze_time("2022-01-01T00:00:00Z"):
        command = ExportAssetsCommand()
        output = [(file[0], file[1]()) for file in list(command.run())]
    assert output == [
        (
            "metadata.yaml",
            "version: 1.0.0\ntype: assets\ntimestamp: '2022-01-01T00:00:00+00:00'\n",
        ),
        ("databases/example.yaml", "<DATABASE CONTENTS>"),
        ("datasets/example/dataset.yaml", "<DATASET CONTENTS>"),
        ("charts/pie.yaml", "<CHART CONTENTS>"),
        ("dashboards/sales.yaml", "<DASHBOARD CONTENTS>"),
        ("queries/example/metric.yaml", "<SAVED QUERY CONTENTS>"),
    ]


@pytest.fixture
def mock_export_tags_command_charts_dashboards(mocker):
    export_tags = mocker.patch("superset.commands.tag.export.ExportTagsCommand")

    def _mock_export(dashboard_ids=None, chart_ids=None):
        if not feature_flag_manager.is_feature_enabled("TAGGING_SYSTEM"):
            return iter([])
        return [
            (
                "tags.yaml",
                lambda: yaml.dump(
                    {
                        "tags": [
                            {
                                "tag_name": "tag_1",
                                "description": "Description for tag_1",
                            }
                        ]
                    },
                    sort_keys=False,
                ),
            ),
            ("charts/pie.yaml", lambda: "tag:\n- tag_1"),
        ]

    export_tags.return_value._export.side_effect = _mock_export
    return export_tags


def test_export_tags_with_charts_dashboards(
    mock_export_tags_command_charts_dashboards, mocker
):
    with patch.object(feature_flag_manager, "is_feature_enabled", return_value=True):
        command = mock_export_tags_command_charts_dashboards()
        result = list(command._export(chart_ids=[1]))

        file_name, file_content_func = result[0]
        file_content = file_content_func()
        assert file_name == "tags.yaml"
        payload = yaml.safe_load(file_content)
        assert payload["tags"] == [
            {"tag_name": "tag_1", "description": "Description for tag_1"}
        ]

        file_name, file_content_func = result[1]
        file_content = file_content_func()
        assert file_name == "charts/pie.yaml"
        assert file_content == "tag:\n- tag_1"

    with patch.object(feature_flag_manager, "is_feature_enabled", return_value=False):
        command = mock_export_tags_command_charts_dashboards()
        result = list(command._export(chart_ids=[1]))
        assert not any(file_name == "tags.yaml" for file_name, _ in result)
        assert all(
            file_content_func() != "tag:\n- tag_1" for _, file_content_func in result
        )

File: superset-frontend/src/utils/export.test.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { SupersetClient } from '@superset-ui/core';
import { logging } from '@apache-superset/core';
import contentDisposition from 'content-disposition';
import handleResourceExport from './export';
import { Mock } from 'vitest';

// Mock dependencies
vi.mock('@superset-ui/core', async (importActual) => ({
  SupersetClient: {
    get: vi.fn(),
  },
}));

vi.mock('@apache-superset/core', () => ({
  logging: {
    warn: vi.fn(),
    error: vi.fn(),
  },
}));

vi.mock('content-disposition');

const { ensureAppRootMock } = vi.hoisted(() => ({
  ensureAppRootMock: vi.fn((path: string) => path),
}));

// Default no-op mock for pathUtils; specific tests customize ensureAppRoot to simulate app root prefixing
vi.mock('./pathUtils', () => ({
  ensureAppRoot: ensureAppRootMock,
}));

let mockBlob: Blob;
let mockResponse: Response;
let createElementSpy: Mock;
let createObjectURLSpy: Mock;
let revokeObjectURLSpy: Mock;

beforeEach(() => {
  // Reset all mocks
  vi.clearAllMocks();

  // Mock Blob
  mockBlob = new Blob(['test data'], { type: 'application/zip' });

  // Mock Response with Headers
  mockResponse = {
    headers: new Headers({
      'Content-Disposition': 'attachment; filename="dashboard_export.zip"',
    }),
    blob: vi.fn().mockResolvedValue(mockBlob),
  } as unknown as Response;

  // Mock SupersetClient.get
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  // Mock DOM APIs
  const mockAnchor = document.createElement('a');
  mockAnchor.click = vi.fn();
  createElementSpy = vi
    .spyOn(document, 'createElement')
    .mockReturnValue(mockAnchor);
  vi.spyOn(document.body, 'appendChild').mockImplementation(() => mockAnchor);
  vi.spyOn(document.body, 'removeChild').mockImplementation(() => mockAnchor);

  // Mock URL.createObjectURL and revokeObjectURL
  createObjectURLSpy = vi
    .spyOn(window.URL, 'createObjectURL')
    .mockReturnValue('blob:mock-url');

  // Create revokeObjectURL if it doesn't exist
  if (!window.URL.revokeObjectURL) {
    window.URL.revokeObjectURL = vi.fn();
  }
  revokeObjectURLSpy = vi.spyOn(window.URL, 'revokeObjectURL');
});

afterEach(() => {
  createElementSpy.mockRestore();
  createObjectURLSpy.mockRestore();
  if (revokeObjectURLSpy) {
    revokeObjectURLSpy.mockRestore();
  }
});

test('exports resource with correct endpoint and headers', async () => {
  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1, 2, 3], doneMock);

  expect(SupersetClient.get).toHaveBeenCalledWith({
    endpoint: '/api/v1/dashboard/export/?q=!(1,2,3)',
    headers: {
      Accept: 'application/zip, application/x-zip-compressed, text/plain',
    },
    parseMethod: 'raw',
  });
});

test('creates blob and triggers download', async () => {
  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  // Check that blob was created
  expect(mockResponse.blob).toHaveBeenCalled();

  // Check that object URL was created
  expect(window.URL.createObjectURL).toHaveBeenCalledWith(mockBlob);

  // Check that anchor element was created and configured
  expect(document.createElement).toHaveBeenCalledWith('a');
  const anchor = document.createElement('a');
  expect(anchor.href).toBe('blob:mock-url');
  expect(anchor.download).toBe('dashboard_export.zip');
  // eslint-disable-next-line jest-dom/prefer-to-have-style -- toHaveStyle not available without React Testing Library
  expect(anchor.style.display).toBe('none');

  // Check that click was triggered
  expect(anchor.click).toHaveBeenCalled();

  // Check cleanup
  expect(document.body.removeChild).toHaveBeenCalled();
  expect(window.URL.revokeObjectURL).toHaveBeenCalledWith('blob:mock-url');
});

test('calls done callback on success', async () => {
  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  expect(doneMock).toHaveBeenCalled();
});

test('uses default filename when Content-Disposition is missing', async () => {
  mockResponse = {
    headers: new Headers(),
    blob: vi.fn().mockResolvedValue(mockBlob),
  } as unknown as Response;
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  const doneMock = vi.fn();
  await handleResourceExport('chart', [42], doneMock);

  const anchor = document.createElement('a');
  expect(anchor.download).toBe('chart_export.zip');
});

test('handles Content-Disposition parsing errors gracefully', async () => {
  (contentDisposition.parse as Mock).mockImplementationOnce(() => {
    throw new Error('Invalid header');
  });

  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  // Should fall back to default filename
  const anchor = document.createElement('a');
  expect(anchor.download).toBe('dashboard_export.zip');
  expect(doneMock).toHaveBeenCalled();
});

test('handles API errors and calls done callback', async () => {
  const apiError = new Error('API Error');
  (SupersetClient.get as Mock).mockRejectedValue(apiError);

  const doneMock = vi.fn();

  await expect(
    handleResourceExport('dashboard', [1], doneMock),
  ).rejects.toThrow('API Error');

  expect(doneMock).toHaveBeenCalled();
});

test('handles blob conversion errors', async () => {
  const blobError = new Error('Blob conversion failed');
  mockResponse.blob = vi.fn().mockRejectedValue(blobError);
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  const doneMock = vi.fn();

  await expect(
    handleResourceExport('dashboard', [1], doneMock),
  ).rejects.toThrow('Blob conversion failed');

  expect(doneMock).toHaveBeenCalled();
});

test('exports multiple resources with correct IDs', async () => {
  const doneMock = vi.fn();
  await handleResourceExport('dataset', [10, 20, 30, 40], doneMock);

  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/dataset/export/?q=!(10,20,30,40)',
    }),
  );
});

test('parses filename from Content-Disposition with quotes', async () => {
  (contentDisposition.parse as Mock).mockReturnValueOnce({
    type: 'attachment',
    parameters: { filename: 'my_custom_export.zip' },
  });

  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  const anchor = document.createElement('a');
  expect(anchor.download).toBe('my_custom_export.zip');
});

test('warns when export exceeds maximum blob size', async () => {
  const largeFileSize = 150 * 1024 * 1024; // 150MB

  mockResponse = {
    headers: new Headers({
      'Content-Length': largeFileSize.toString(),
      'Content-Disposition': 'attachment; filename="large_export.zip"',
    }),
    blob: vi.fn().mockResolvedValue(mockBlob),
  } as unknown as Response;
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  expect(logging.warn).toHaveBeenCalledWith(
    expect.stringContaining('exceeds maximum blob size'),
  );
  expect(doneMock).toHaveBeenCalled();
});

test('handles various resource types', async () => {
  const doneMock = vi.fn();

  await handleResourceExport('dashboard', [1], doneMock);
  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/dashboard/export/?q=!(1)',
    }),
  );

  await handleResourceExport('chart', [1], doneMock);
  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/chart/export/?q=!(1)',
    }),
  );

  await handleResourceExport('dataset', [1], doneMock);
  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/dataset/export/?q=!(1)',
    }),
  );

  await handleResourceExport('database', [1], doneMock);
  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/database/export/?q=!(1)',
    }),
  );

  await handleResourceExport('query', [1], doneMock);
  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/query/export/?q=!(1)',
    }),
  );

  expect(doneMock).toHaveBeenCalledTimes(5);
});

test('handles network errors and logs them', async () => {
  const networkError = new Error('Network request failed');
  (SupersetClient.get as Mock).mockRejectedValue(networkError);

  const doneMock = vi.fn();

  await expect(
    handleResourceExport('dashboard', [1], doneMock),
  ).rejects.toThrow('Network request failed');

  expect(logging.error).toHaveBeenCalledWith(
    'Resource export failed:',
    networkError,
  );
  expect(doneMock).toHaveBeenCalled();
});

test('handles 404 errors when resource not found', async () => {
  const notFoundError = new Error('Not found');
  (SupersetClient.get as Mock).mockRejectedValue(notFoundError);

  const doneMock = vi.fn();

  await expect(
    handleResourceExport('dashboard', [999], doneMock),
  ).rejects.toThrow('Not found');

  expect(doneMock).toHaveBeenCalled();
});

test('handles empty response from server', async () => {
  const emptyBlob = new Blob([], { type: 'application/zip' });
  mockResponse = {
    headers: new Headers({
      'Content-Disposition': 'attachment; filename="empty.zip"',
    }),
    blob: vi.fn().mockResolvedValue(emptyBlob),
  } as unknown as Response;
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [1], doneMock);

  expect(window.URL.createObjectURL).toHaveBeenCalledWith(emptyBlob);
  expect(doneMock).toHaveBeenCalled();
});

test('cleans up blob URL even when download fails', async () => {
  const mockAnchor = document.createElement('a');
  mockAnchor.click = vi.fn().mockImplementation(() => {
    throw new Error('Click failed');
  });

  createElementSpy.mockRestore();
  createElementSpy = vi
    .spyOn(document, 'createElement')
    .mockReturnValue(mockAnchor);

  const doneMock = vi.fn();

  await expect(
    handleResourceExport('dashboard', [1], doneMock),
  ).rejects.toThrow('Click failed');

  // Verify cleanup still happens
  expect(window.URL.revokeObjectURL).toHaveBeenCalledWith('blob:mock-url');
  expect(doneMock).toHaveBeenCalled();
});

test('handles malformed Content-Disposition header', async () => {
  mockResponse = {
    headers: new Headers({
      'Content-Disposition': 'not-a-valid-header',
    }),
    blob: vi.fn().mockResolvedValue(mockBlob),
  } as unknown as Response;
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  (contentDisposition.parse as Mock).mockImplementationOnce(() => {
    throw new Error('Parse error');
  });

  const doneMock = vi.fn();
  await handleResourceExport('dataset', [5], doneMock);

  // Should fall back to default filename
  const anchor = document.createElement('a');
  expect(anchor.download).toBe('dataset_export.zip');
  expect(logging.warn).toHaveBeenCalledWith(
    'Failed to parse Content-Disposition header:',
    expect.any(Error),
  );
});

test('handles missing headers object', async () => {
  mockResponse = {
    headers: new Headers(),
    blob: vi.fn().mockResolvedValue(mockBlob),
  } as unknown as Response;
  (SupersetClient.get as Mock).mockResolvedValue(mockResponse);

  const doneMock = vi.fn();
  await handleResourceExport('chart', [7], doneMock);

  const anchor = document.createElement('a');
  expect(anchor.download).toBe('chart_export.zip');
  expect(doneMock).toHaveBeenCalled();
});

test('handles export with empty IDs array', async () => {
  const doneMock = vi.fn();
  await handleResourceExport('dashboard', [], doneMock);

  expect(SupersetClient.get).toHaveBeenCalledWith(
    expect.objectContaining({
      endpoint: '/api/v1/dashboard/export/?q=!()',
    }),
  );
});

const doublePrefixTestCases = [
  {
    name: 'subdirectory prefix',
    appRoot: '/superset',
    resource: 'dashboard',
    ids: [1],
  },
  {
    name: 'subdirectory prefix (dataset)',
    appRoot: '/superset',
    resource: 'dataset',
    ids: [1],
  },
  {
    name: 'nested prefix',
    appRoot: '/my-app/superset',
    resource: 'dataset',
    ids: [1, 2],
  },
];

test.each(doublePrefixTestCases)(
  'handleResourceExport endpoint should not include app prefix: $name',
  async ({ appRoot, resource, ids }) => {
    // Simulate real ensureAppRoot behavior: prepend the appRoot
    ensureAppRootMock.mockImplementation((path: string) => `${appRoot}${path}`);

    const doneMock = vi.fn();
    await handleResourceExport(resource, ids, doneMock);

    // The endpoint passed to SupersetClient.get should NOT have the appRoot prefix
    // because SupersetClient.getUrl() adds it when building the full URL.
    const expectedEndpoint = `/api/v1/${resource}/export/?q=!(${ids.join(',')})`;

    // Explicitly verify no prefix in endpoint - this will fail if ensureAppRoot is used
    const callArgs = (SupersetClient.get as Mock).mock.calls.slice(-1)[0][0];
    expect(callArgs.endpoint).not.toContain(appRoot);
    expect(callArgs.endpoint).toBe(expectedEndpoint);

    // Reset mock for next test
    ensureAppRootMock.mockImplementation((path: string) => path);
  },
);
